File: fallbacks.py
"""
Rishta Aunty - Deterministic Fallback & Rule-Based Roast Matrix (Python)
Packed with maximum authentic Pakistani & Desi slangs, witty burns, and matrimonial banter.
"""

RISHTA_FALLBACKS = {
    # Edge Case: Absolute Zero Repos
    "zero_repos": {
        "title": "M.Sc Ghost Developer (Khali Haath Phannay Khan)",
        "gotra": "Incognito Topi Drama Qabeela",
        "habits": [
            "Din bhar LinkedIn aur Twitter par lambi lambi chussain maarta hai",
            "Ghar walon ko jhoot bola hua hai ke 'Silicon Valley stealth startup' chala raha hoon",
            "README generator se bhi pehle ChatGPT ke aage ro deta hai"
        ],
        "red_flags": [
            "GitHub profile aisi saaf hai jaise raid parne ke baad office!",
            "Portfolio ke naam par 4 saal se 'Under Construction' ka board latkaya hua hai",
            "Zero repos, zero commits, par attitude aisa jaise Elon Musk ka secret chacha ho"
        ],
        "aunty_verdict": "Haye tauba lakh di laanat! Zero public repos?! Beta, khali biodata le kar rishta dhoondne nikal pare ho? Ladki walay ghar aayein aur poochhein 'damad ji kya karte hain', toh kya bolenge 'bhai inspect element mein hoshiyari dikhata hai'? Khuda ka khauf karo, pehle ek chhota sa 'Hello World' toh push kar aao, warna Phupho ne pehle round mein hi rishte ka scene off kar dena hai!",
        "match_score": 9,
        "stamp_status": "SCENE OFF HAI ❌ (RISHTA KHARIJ)",
        "stamp_class": "stamp-rejected"
    },

    # Edge Case: 1 to 3 Repos (Barely Started / Tutorial Inmate)
    "novice_repos": {
        "title": "Pappu Tutorial Inmate (Abhi Chhota Bacha Hai)",
        "gotra": "10-Minute Udemy Shaukeen Baradri",
        "habits": [
            "Ek Todo app aur ek Calculator bana kar mohallay mein shashkay maar raha hai",
            "Commit message mein 'first commit', 'final fix', 'final fix 2 pakka' likhta hai",
            "Ammi ko kehta hai 'agla Bill Gates main hi hoon, bas 500 rupay ka easypaisa kar dein'"
        ],
        "red_flags": [
            "YouTube tutorial pause kar karke line-by-line copy paste marta hai",
            "Git push karte waqt aisi duayein mangta hai jaise matric ka result aa raha ho",
            "Div ko center karne ke chakkar mein pura hafta depression mein guzaar deta hai"
        ],
        "aunty_verdict": "Aray beta ji, do repo bana kar rishte ke bazaar mein itna roab? Ek Todo list aur ek silly calculator se biwi ke nakhray uthao ge? Magar chalo, kam az kam larka koshish toh kar raha hai. Ladki walon ko bolenge 'abhi junior level ka jugaadu hai, shaadi ke baad thoda aur polish ho jayega'!",
        "match_score": 35,
        "stamp_status": "PHUPHO KO SAKHT AITRAZ ⚠️",
        "stamp_class": "stamp-pending"
    },

    # Edge Case: Fork Hoarder (> 50% repos are forks)
    "fork_hoarder": {
        "title": "Master of Doosron Ka Code (Open Source Chor / Jugaad King)",
        "gotra": "Ctrl+C Ctrl+V Khandaan",
        "habits": [
            "Trending tab dekhte hi har kisi ka repo bina sharam ke fork maar leta hai",
            "Doosron ka project fork karke bas README mein apna naam chipka deta hai",
            "Bio mein likha hai 'Core Contributor' (bas ek comma theek kiya tha)"
        ],
        "red_flags": [
            "Apna original code dhoondne ke liye forensic audit karwana parega",
            "Shaadi mein dahej bhi parosiyon ke ghar se mang kar show-off karega",
            "Har baat pe mamoo banane ki aadat hai, code mein bhi aur rishtey mein bhi"
        ],
        "aunty_verdict": "Tauba tauba aisi chori chakari! Itnay saare forked repos?! Beta, doosron ki mehnat ko apna rishta biodata bana kar shashkay dikha rahe ho? Yeh toh bilkul aisi baat hui ke parosi ki Mercedes ke aage kharay ho kar rishte ki tasveer khinchwa li! Original code kidhar hai jani? Abba nahi manenge aisi topi drama par!",
        "match_score": 26,
        "stamp_status": "TOPI DRAMA EXPOSED ❌",
        "stamp_class": "stamp-rejected"
    },

    # Archetype: JavaScript / TypeScript Overloader
    "javascript_addict": {
        "title": "Senior Framework Hopper (npm install ka Nashai)",
        "gotra": "Bhari Bhari Node_Modules Biradari",
        "habits": [
            "Har mangalwar naya JS framework dhoond ke purana code kachre mein phenk deta hai",
            "Node_modules itna heavy hai ke hard drive cheekhein maar rahi hai",
            "Shaadi ke nikaah naame par bhi TypeScript ke strict types declare karne ki zid karega"
        ],
        "red_flags": [
            "Commitment issues peak level par hain! Aaj React, kal Next.js, parso Remix, kahin tikta hi nahi",
            "UI mein dark mode aur glowing buttons toh daal diye, par login button dabao toh 500 error deta hai",
            "Chai garam karne jao toh bolta hai 'Bun is 10x faster than traditional kettle, bro'"
        ],
        "aunty_verdict": "Haye mera BP! Beta dimagh ka tez hai par iske commitment par shaq hai! Aaj keh raha hai React pasand hai, kal bolega Svelte pe dil aa gaya. Kal ko biwi se kahega 'jaan, humara rishta thoda lag kar raha hai, let me refactor our marriage into microservices'. Thoda thehrav lao zindagi mein beta, har cheez npm update se solve nahi hoti!",
        "match_score": 48,
        "stamp_status": "THODA SOCH VICHAR BAKI ⚠️",
        "stamp_class": "stamp-pending"
    },

    # Archetype: Python / AI Enthusiast
    "python_ai_fan": {
        "title": "Self-Styled AI Guru (Prompt Engineer Phannay Khan)",
        "gotra": "Jupyter Notebook & GPU Frying Jamat",
        "habits": [
            "Din raat `import torch` aur `import openai` ke khwaab dekhta rehta hai",
            "Rishta proposal aur love letter likhne ke liye bhi Claude 3.5 Sonnet ka sahara liya hai",
            "Mining rig aur GPU heating par roti garam karne ka jugaad lagata hai"
        ],
        "red_flags": [
            "Code run karne jao toh pehle 8 GB weights aur huggingface models download karta hai",
            "Agar internet band ho jaye toh ek simple 'for loop' likhte huye pasine chhoot jate hain",
            "Har conversation mein 'AI will replace humans' bol ke rishtedaaron ko bore karta hai"
        ],
        "aunty_verdict": "Wah bhai wah! Beta AI ka bada ustaad banta hai! Bio mein likha hai 'Machine Learning Visionary', par saari files `final_notebook_real_final_v3_copy.ipynb` hain! Chalo kam az kam shaadi ke baad sasural walon ki puraani tasveerein 4K mein enhance karke de dega. Biryani mein elaichi jaisa thoda ajeeb hai, par kamaai theek thak kar lega!",
        "match_score": 63,
        "stamp_status": "RISHTA UNDER INVESTIGATION 📜",
        "stamp_class": "stamp-pending"
    },

    # Archetype: C++ / Rust / Systems Purist
    "systems_purist": {
        "title": "Memory Leak Inspector & Low-Level Aristocrat",
        "gotra": "Borrow Checker & Segmentation Fault Biradari",
        "habits": [
            "Har baat par be-tukka dialogue maarta hai: 'Rust is memory safe, unlike our society'",
            "Shaadi ke agreement ko compile-time verify karne ke chakkar mein dawat miss kar deta hai",
            "Subah 5 bajay uth kar terminal ke NeoVim plugins customize karta rehta hai"
        ],
        "red_flags": [
            "Biwi paani maang le toh pointer reference pass kar dega",
            "Zero social skills! Khaandani shaadiyon mein baith kar logon ko Linux install karne ke bhashan deta hai",
            "Zindagi bhar Garbage Collector ke khilaf jahaad karta rahega"
        ],
        "aunty_verdict": "MashaAllah dimaag toh Einstein jaisa hai, par dil kahan hai beta? Is se rishte ki baat karo toh aage se kehta hai 'lifetime annotations mismatch'. Arre bhai, hum yahan lifetime partner dhoond rahe hain, koi compiler error nahi! Par chalo, package acha hai aur larka sharif hai, bas romantic dates par terminal band rakhne ka waada lena parega!",
        "match_score": 72,
        "stamp_status": "KHANDAANI APPROVED RISHTA 💍",
        "stamp_class": "stamp-verified"
    },

    # Archetype: Abandoned Projects Specialist (Lots of repos, 0 stars)
    "abandoned_repo_king": {
        "title": "Cemetery of Incomplete Startups (Khwaboon Ka Saudagar)",
        "gotra": "Domain Name Kharidar & Abandoner Sect",
        "habits": [
            "Har doosre jummay naya .com domain khareedta hai, par site kabhi live nahi hoti",
            "47 repos banaye hain, 46 par sirf initial commit aur ek tuta hua README hai",
            "Har kisi ko kehta hai 'bhai market ready nahi thi mere revolutionary product ke liye'"
        ],
        "red_flags": [
            "Starting things: 1000% Josh. Finishing things: Zero battey sannaata!",
            "Ghar ka pankha kharab ho jaye toh 6 mahine tak GitHub issue open rakhta hai",
            "Commit streak aisi hai jaise Karachi mein bijli ki supply (kabhi aayi, kabhi gayi)"
        ],
        "aunty_verdict": "Beta has 47 repos, 46 abandoned! Shuru bohot josh-o-kharoosh se karta hai, phir do din baad hawa nikal jaati hai! Kal ko shaadi karke do haftay baad bolega 'I have pivoted to bachelor life again'. Pehle ek project mukammal karke production par deploy karo jani, phir kisi masoom bachi ka haath mangne aana!",
        "match_score": 28,
        "stamp_status": "COMMITMENT ISSUES PEAK PAR ❌",
        "stamp_class": "stamp-rejected"
    },

    # Archetype: Open Source Superstar (Sharma Ji Ka Beta)
    "open_source_chad": {
        "title": "Sharma Ji Ka Beta (Full Shashkay Coder)",
        "gotra": "Top 1% Stargazer & Green Graph Royals",
        "habits": [
            "Subah nashte mein paratha chabaate chabaate PR merge karta hai",
            "Followers itnay hain ke mohallay ka MPA election aaram se jeet jaye",
            "Har tech conference mein speaker wala lanyard pehan ke hero ban ke ghoomta hai"
        ],
        "red_flags": [
            "Bohot zyada mashroof! Honeymoon par bhi laptop khol ke issues triage karega",
            "Nakhray bohot honge, biwi ki baat sunne ke liye issue template submit karne ko kahega",
            "Shaadi ke shor-sharabay mein bhi commit streak tootte nahi dega"
        ],
        "aunty_verdict": "SubhanAllah, MashaAllah! Hazaron stars aur itnay followers! Yeh toh sach mein Sharma ji ka beta nikla! Mohallay ki saari auntiyan iske ghar rishtey le kar pohnchi hui hain. Paisa bhi khoob kama raha hai aur khandaan ki naak bhi oonchi kar di. Larka pakka heera hai, rishta abhi lock karo warna koi aur aunty le uregi!",
        "match_score": 93,
        "stamp_status": "RISHTA 100% PAKKA! 💖",
        "stamp_class": "stamp-verified"
    },

    # Archetype: Sleep-Deprived Night Owl
    "workaholic_vampire": {
        "title": "Raat Ka Coder (Neend Ka Dushman / RedBull Damad)",
        "gotra": "3 AM Deployments & Kali Chai Jamat",
        "habits": [
            "Dopehar 2 bajay aankh khulti hai, raat 3 bajay terminal aag ugalta hai",
            "Suraj ki roshni dekh kar aisi cheekh maarta hai jaise Dracula ho",
            "Sasural walon ki dawat par kone mein baith kar laptop par hotfix pel raha hota hai"
        ],
        "red_flags": [
            "Aankhon ke neeche dark circles itnay hain jaise do black holes ban gaye hon",
            "Subah nashta table par so raha hoga aur raat ko chai ki kettliyan khali karega",
            "Romantic long drive par bhi kahega 'hotspot on karo, prod fat gaya hai'"
        ],
        "aunty_verdict": "Beta ji, raat ke teen bajay kaun production par code push karta hai? Biwi kahegi 'chalo dawat pe chalte hain', yeh aage se bolega 'ek prod incident aag lagaye baitha hai'. Magar dil ka bura nahi hai, mehnat kash banda hai aur biwi ko designer jore zaroor dilaayega. Bas shaadi ke baad timing seedhi karni paregi!",
        "match_score": 58,
        "stamp_status": "CONDITIONAL APPROVAL ⚠️",
        "stamp_class": "stamp-pending"
    },

    # Archetype: Solid Balanced Dev
    "solid_developer": {
        "title": "Sharif Khandani Developer (Sadiq & Ameen Coder)",
        "gotra": "Clean Code & Proper Documentation Jamat",
        "habits": [
            "README theek se likhta hai, tests bhi pass karwata hai (kabhi kabhi)",
            "Office se waqt par nikalta hai, family ko time dene ki poori salahiyat hai",
            "Chai time par thanda hone se pehle table par aa jata hai"
        ],
        "red_flags": [
            "Thoda seedha-saadha aur boring hai, weekends par documentation padhta hai",
            "Chai mein cheeni kam ho toh thoda sa munh phula leta hai",
            "Har mehmaan ko zabardasti Linux ke faiday sunata hai"
        ],
        "aunty_verdict": "MashaAllah, bohot hi suljha hua aur shareef bacha hai! Code bhi saaf suthra hai aur harkatein bhi khandani. Na koi fazool shashkay, na jhooti hawa-baazi. Bas kabhi kabhar Git rebase mein phans kar thoda ghabra jata hai, par humari taraf se rishta poora approve hai!",
        "match_score": 82,
        "stamp_status": "AUNTY APPROVED RISHTA ✅",
        "stamp_class": "stamp-verified"
    }
}


def generate_fallback_roast(profile: dict, repos: list = None) -> dict:
    """
    Intelligent Rule-Matcher in Python with maximum Desi & Pakistani slangs.
    """
    if repos is None:
        repos = []

    public_repos = profile.get("public_repos", 0) or 0
    followers = profile.get("followers", 0) or 0
    following = profile.get("following", 0) or 0
    stars = sum((r.get("stargazers_count", 0) or 0) for r in repos)
    forks_count = sum(1 for r in repos if r.get("fork"))
    fork_ratio = (forks_count / len(repos)) if repos else 0.0

    # Language distribution
    lang_counts = {}
    for r in repos:
        lang = r.get("language")
        if lang:
            lang_counts[lang] = lang_counts.get(lang, 0) + 1

    top_lang = "Plain Text"
    if lang_counts:
        top_lang = sorted(lang_counts.items(), key=lambda item: item[1], reverse=True)[0][0]

    # Matching logic
    if public_repos == 0:
        base = RISHTA_FALLBACKS["zero_repos"]
    elif public_repos <= 2:
        base = RISHTA_FALLBACKS["novice_repos"]
    elif fork_ratio >= 0.55 and len(repos) >= 4:
        base = RISHTA_FALLBACKS["fork_hoarder"]
    elif stars > 250 or followers > 300:
        base = RISHTA_FALLBACKS["open_source_chad"]
    elif top_lang in ["JavaScript", "TypeScript"] and public_repos > 12:
        base = RISHTA_FALLBACKS["javascript_addict"]
    elif top_lang in ["Python", "Jupyter Notebook"]:
        base = RISHTA_FALLBACKS["python_ai_fan"]
    elif top_lang in ["Rust", "C++", "C", "Go"]:
        base = RISHTA_FALLBACKS["systems_purist"]
    elif public_repos >= 20 and stars < 5:
        base = RISHTA_FALLBACKS["abandoned_repo_king"]
    elif followers < 5 and following > 80:
        base = RISHTA_FALLBACKS["workaholic_vampire"]
    else:
        base = RISHTA_FALLBACKS["solid_developer"]

    import copy
    customized = copy.deepcopy(base)

    candidate_name = profile.get("name") or profile.get("login") or "Beta"
    customized["candidateName"] = candidate_name
    customized["topLanguage"] = top_lang
    customized["totalStars"] = stars
    customized["publicRepos"] = public_repos
    customized["followers"] = followers
    customized["forkRatioPercent"] = round(fork_ratio * 100)

    # Slang-packed Assets list
    if public_repos > 0:
        customized["assets"] = [
            f"{public_repos} Public Repos (Dahej ka total saamaan)",
            f"{stars} Total Stars (Mohallay mein izzat aur shashkay)",
            f"{followers} Followers vs {following} Following ({'Mashoor Celebrity Coder' if followers > following else 'Sab ko follow karta phirta hai bechara'})",
            f"Kamaai Ka Jugaad: {top_lang} ({'Full nakhray baaz tech' if top_lang == 'JavaScript' else 'Mehnati majdoor coder'})"
        ]
    else:
        customized["assets"] = [
            "0 Public Repos (Ghar ki almaari bilkul khali hai)",
            "0 Stars (Mohallay ke bachon ne bhi like nahi kiya)",
            "Private repos ka bahana (Kehta hai confidential scene hai)",
            "Zero verifiable hunar (Bas baatein hi baatein)"
        ]

    # Adjust score slightly for uniqueness
    bonus = min(15, (stars // 20) + (followers // 50))
    score = base["match_score"] + (bonus if stars > 10 else -bonus)
    customized["match_score"] = min(98, max(8, score))

    return {
        "title": customized["title"],
        "gotra": customized["gotra"],
        "habits": customized["habits"],
        "assets": customized["assets"],
        "redFlags": customized["red_flags"],
        "auntyVerdict": customized["aunty_verdict"],
        "matchScore": customized["match_score"],
        "stampStatus": customized["stamp_status"],
        "stampClass": customized["stamp_class"],
        "candidateName": customized["candidateName"],
        "topLanguage": customized["topLanguage"],
        "totalStars": customized["totalStars"],
        "publicRepos": customized["publicRepos"],
        "followers": customized["followers"]
    }

File: fallbacks.js
/**
 * Rishta Aunty - Deterministic Fallback & Rule-Based Roast Matrix
 * 
 * Guarantees zero-latency, culturally hilarious, and 100% reliable biodata
 * roasts during live stage demos, even if offline or rate-limited.
 */

const RISHTA_FALLBACKS = {
  // Edge Case: Absolute Zero Repos
  zeroRepos: {
    title: "M.Sc Ghost Developer (Khali Haath)",
    gotra: "Incognito Committer Qabeela",
    habits: [
      "Din bhar Twitter aur LinkedIn par tech debate karta hai",
      "README bhi generate karne ke liye ChatGPT ka intezar karta hai",
      "Ghar walon ko bola hai 'stealth startup' chala raha hai"
    ],
    redFlags: [
      "Profile par ek bhi green tile nahi hai (Dil bhi khali hai)",
      "Portfolio website ke naam par 'Coming Soon' laga rakha hai",
      "Zero commits, 100% confidence"
    ],
    auntyVerdict: "Haye tauba! Zero public repos?! Beta, khali biodata le kar rishta dhoondne nikal pare ho? Ladki walon ne poochha 'beta kya karta hai', toh kya bolenge 'inspect element karta hai'? Pehle ek chhota sa 'Hello World' toh push kar lo, phir hum kisi ko chai par bulayein!",
    matchScore: 12,
    stampStatus: "RISHTA KHARIJ ❌",
    stampClass: "stamp-rejected"
  },

  // Edge Case: 1 to 3 Repos (Barely Started)
  noviceRepos: {
    title: "Junior Tutorial Inmate (Abhi seekh raha hai)",
    gotra: "Udemy Certified Sect",
    habits: [
      "Todo app bana kar khud ko Full Stack architect samajhta hai",
      "Commit message mein sirf 'first commit' aur 'fixed bug' likhta hai",
      "Ammi ko kehta hai 'Agla Sundar Pichai main hi hoon'"
    ],
    redFlags: [
      "Saare projects YouTube video dekh kar line-by-line copy kiye hain",
      "Git push ke baad dua maangta hai ke conflict na aaye",
      "CSS align center karne ke liye 3 ghante lagata hai"
    ],
    auntyVerdict: "Aray beta ji, do repo bana kar rishte ke bazaar mein entry marli? Ek Todo App aur ek Calculator se ghar nahi chalta! Magar chalo, shauk toh hai bachay ko. Ladki walon ko bolenge 'thoda sa fixing baki hai, update ho jayega'.",
    matchScore: 38,
    stampStatus: "LADKI WALAY SOCHENGAY ⚠️",
    stampClass: "stamp-pending"
  },

  // Edge Case: Fork Hoarder (> 50% repos are forks)
  forkHoarder: {
    title: "Master of Doosron Ka Code (Open Source Chor)",
    gotra: "Ctrl+C Ctrl+V Khandaan",
    habits: [
      "Har trending library ko bina soche fork kar leta hai",
      "Fork karke star dena bhool jata hai",
      "Bio mein 'Contributor to Kubernetes' likha hai (bas typo fix kiya tha)"
    ],
    redFlags: [
      "Apna original code do line ka nahi hai, baki sab borrowed glory",
      "Rishte mein dahej bhi doosron se maang kar dikhayega",
      "Git clone karke file rename karke upload karta hai"
    ],
    auntyVerdict: "Tauba tauba! Itnay saare forked repos?! Beta, doosron ki mehnat ko apna rishta biodata bana kar pesh kar rahe ho? Yeh toh aisi baat hui ke parosi ki gaari ke aage kharay ho kar tasveer khinchwa li! Original kaam kahan hai?",
    matchScore: 29,
    stampStatus: "RISHTA KHARIJ ❌",
    stampClass: "stamp-rejected"
  },

  // Archetype: JavaScript / TypeScript Overloader
  javascriptAddict: {
    title: "Senior Framework Hopper (npm install ka badshah)",
    gotra: "Node_Modules Bhari Gotra",
    habits: [
      "Har naye hafte naya JS framework seekhne lag jata hai",
      "Node_modules folder itna bhaari hai ke hard drive cheekhein maar rahi hai",
      "Shaadi ke card par bhi TypeScript ke types declare karega"
    ],
    redFlags: [
      "Commitment issues hain! React se Next.js, Next se Remix, kahin tikta hi nahi",
      "Dark mode toh add kar liya, par functionality aadhi tooti hui hai",
      "Chai garam karne ke bajaye 'Bun is faster than Node' par larta hai"
    ],
    auntyVerdict: "Beta bohot tez hai, par iske commitment par shaq hai! Kal React pasand tha, aaj Svelte pe fida hai, parso kisi aur pe chala jayega. Biwi se bolay ga 'jaan, let's refactor our marriage into microservices'. Thoda thehrav lao beta zindagi mein!",
    matchScore: 47,
    stampStatus: "CONDITIONAL MATCH ⚠️",
    stampClass: "stamp-pending"
  },

  // Archetype: Python / AI Enthusiast
  pythonAIFan: {
    title: "Self-Styled AI Guru (Prompt Engineer Sahib)",
    gotra: "Jupyter Notebook Caste",
    habits: [
      "Pura din `import torch` aur `import openai` ke khwaab dekhta hai",
      "Rishta proposal likhne ke liye bhi Claude aur ChatGPT use kiya hai",
      "GPU heating par roti garam karne ki koshish karta hai"
    ],
    redFlags: [
      "Code chalne se pehle 4 GB weights download karta hai",
      "Bina LLM ke ek for loop nahi likh sakta",
      "Ghar ka AC band karwa ke mining rig chala rakhi hai"
    ],
    auntyVerdict: "Wah bhai wah! Beta AI ka shauqeen hai. Bio mein likha hai 'Machine Learning Engineer', par saari files `test_final_v2_really_final.ipynb` hain! Chalo kam az kam shaadi ke baad sasural walon ki tasveerein toh enhance karke de dega.",
    matchScore: 61,
    stampStatus: "RISHTA UNDER REVIEW 📜",
    stampClass: "stamp-pending"
  },

  // Archetype: C++ / Rust / Systems Dev
  systemsPurist: {
    title: "Memory Leak Inspector & Borrow Checker Devotee",
    gotra: "Low-Level Aristocracy",
    habits: [
      "Baat baat par bolta hai 'Rust is memory safe, unlike our society'",
      "Shaadi ke contracts ko compile-time check karna chahta hai",
      "Subah 5 bajay uth kar terminal ke font customize karta hai"
    ],
    redFlags: [
      "Biwi ne paani maanga toh pointer reference de dega",
      "Zero social skills, sirf StackOverflow par logon ko galat sabit karta hai",
      "Zindagi bhar garbage collection ke khilaf larta rahega"
    ],
    auntyVerdict: "MashaAllah dimaag toh bohot tez hai, par dil kahan hai? Is se baat karo toh kehta hai 'lifetime annotations mismatch'. Arre beta, hum yahan lifetime partner dhoond rahe hain, compiler warnings nahi! Par chalo package acha mil jayega.",
    matchScore: 68,
    stampStatus: "KHANDAANI PASAND 💍",
    stampClass: "stamp-verified"
  },

  // Archetype: Abandoned Projects Specialist (Lots of repos, 0 stars, low activity)
  abandonedRepoKing: {
    title: "Cemetery of Incomplete Startups Founder",
    gotra: "Domain Name Kharidar Biradari",
    habits: [
      "Har doosre mahine naya domain khareedta hai, par site kabhi launch nahi hoti",
      "Har repo mein sirf initial commit aur .gitignore hota hai",
      "Kehta hai 'bhai market ready nahi thi is product ke liye'"
    ],
    redFlags: [
      "Starting things: 100%. Finishing things: 0%",
      "Ghar ka bulb kharab ho toh 4 mahine tak issue open rakhega",
      "Commit streak aisi hai jaise Eid ka chaand"
    ],
    auntyVerdict: "Beta has 47 repos, 46 abandoned! Shuru bohot josh se karta hai, phir do din baad bhool jata hai. Kal ko shaadi karke bolega 'I have pivoted to single life'. Pehle ek project pura karke production pe daalo, phir shaadi ke mandap/nikah mein aana!",
    matchScore: 33,
    stampStatus: "RISHTA KHARIJ ❌",
    stampClass: "stamp-rejected"
  },

  // Archetype: High Star / High Follower Open Source Star
  openSourceChad: {
    title: "Sharma Ji Ka Beta (GitHub Sitarah)",
    gotra: "Top 1% Stargazer Royals",
    habits: [
      "Chai peete peete pull request merge karta hai",
      "Followers itnay hain ke mohallay ka council election jeet jaye",
      "Har conference mein speaker badge pehan ke ghoomta hai"
    ],
    redFlags: [
      "Bohot zyada busy! Honeymoon par bhi GitHub issues triage karega",
      "Nakhray bohot honge, biwi ki baat sunne ke liye issue template mangega",
      "Shaadi ke shor mein bhi commit streak tootte nahi dega"
    ],
    auntyVerdict: "SubhanAllah! 1000+ stars aur itnay followers! Yeh toh Sharma ji ka beta nikla! Saari auntyan iske peeche pari hain. Magar daal mein kuch kaala toh nahi? Itna online rehta hai toh rishte walon ko waqt kab dega? Chalo rishta pakka karte hain, ladka kamao hai!",
    matchScore: 89,
    stampStatus: "RISHTA PAKKA! 💖",
    stampClass: "stamp-verified"
  },

  // Archetype: Workaholic Ghost (Commits at 3 AM / Weekends)
  workaholicVampire: {
    title: "Raat Ka Coder (Sleep Deprived Damad)",
    gotra: "Chai & RedBull Devotees",
    habits: [
      "Subah 11 bajay aankh khulti hai, raat 3 bajay terminal garam hota hai",
      "Sunlight dekh kar aisi aankein band karta hai jaise Dracula ho",
      "Dawat par baith kar laptop par hotfix deploy karta hai"
    ],
    redFlags: [
      "BP high rehta hai, ankhon ke neeche dark circles permanent hain",
      "Sasural walay subah nashta karein toh yeh so raha hoga",
      "Romantic dates par Wi-Fi hotspot dhoondta phirega"
    ],
    auntyVerdict: "Beta ji, raat ke teen bajay kaun code push karta hai? Biwi kahegi 'chalo dawat pe chalte hain', yeh bolega 'ek prod incident solve karke aaya'. Magar mehnti bohot hai, biwi ko shopping zaroor karwayega!",
    matchScore: 54,
    stampStatus: "SOCH BICHAAR JARI ⚠️",
    stampClass: "stamp-pending"
  },

  // Archetype: General Solid Mid-Level Dev
  solidDeveloper: {
    title: "Sharif Khandani Developer (Sadiq & Ameen Coder)",
    gotra: "Clean Code Jamat",
    habits: [
      "README theek se likhta hai, documentation bhi achi hai",
      "Tests run karta hai (kabhi kabhi pass bhi hote hain)",
      "Office se waqt par nikalta hai, family man banne ki salahiyat hai"
    ],
    redFlags: [
      "Thoda boring hai, weekends par documentation padhta hai",
      "Chai thandi hone par thoda chirchira ho jata hai",
      "Linux install karne ki advice har mehfil mein deta hai"
    ],
    auntyVerdict: "MashaAllah bohot suljha hua bacha hai. Code bhi saaf hai aur harkatein bhi sharifana. Na koi fazool show-off, na empty promises. Bas kabhi kabhi Git rebase mein phans jata hai, par hum rishta approve karte hain!",
    matchScore: 78,
    stampStatus: "RISHTA APPROVED ✅",
    stampClass: "stamp-verified"
  }
};

/**
 * Intelligent Rule-Matcher that inspects actual profile metrics
 * and selects the funniest and most accurate fallback roast.
 */
function generateFallbackRoast(profile, repos = []) {
  const publicRepos = profile.public_repos || 0;
  const followers = profile.followers || 0;
  const following = profile.following || 0;
  const stars = repos.reduce((acc, r) => acc + (r.stargazers_count || 0), 0);
  const forksCount = repos.filter(r => r.fork).length;
  const forkRatio = repos.length > 0 ? (forksCount / repos.length) : 0;
  
  // Calculate language breakdown
  const langCounts = {};
  repos.forEach(r => {
    if (r.language) {
      langCounts[r.language] = (langCounts[r.language] || 0) + 1;
    }
  });
  const topLang = Object.keys(langCounts).sort((a, b) => langCounts[b] - langCounts[a])[0] || 'Plain Text';

  // Determine candidate match
  let baseTemplate;

  if (publicRepos === 0) {
    baseTemplate = RISHTA_FALLBACKS.zeroRepos;
  } else if (publicRepos <= 2) {
    baseTemplate = RISHTA_FALLBACKS.noviceRepos;
  } else if (forkRatio >= 0.55 && repos.length >= 4) {
    baseTemplate = RISHTA_FALLBACKS.forkHoarder;
  } else if (stars > 250 || followers > 300) {
    baseTemplate = RISHTA_FALLBACKS.openSourceChad;
  } else if (['JavaScript', 'TypeScript'].includes(topLang) && publicRepos > 12) {
    baseTemplate = RISHTA_FALLBACKS.javascriptAddict;
  } else if (['Python', 'Jupyter Notebook'].includes(topLang)) {
    baseTemplate = RISHTA_FALLBACKS.pythonAIFan;
  } else if (['Rust', 'C++', 'C', 'Go'].includes(topLang)) {
    baseTemplate = RISHTA_FALLBACKS.systemsPurist;
  } else if (publicRepos >= 20 && stars < 5) {
    baseTemplate = RISHTA_FALLBACKS.abandonedRepoKing;
  } else if (followers < 5 && following > 80) {
    baseTemplate = RISHTA_FALLBACKS.workaholicVampire;
  } else {
    baseTemplate = RISHTA_FALLBACKS.solidDeveloper;
  }

  // Deep clone to inject customized stats
  const customized = JSON.parse(JSON.stringify(baseTemplate));

  // Personalize the verdict with their actual numbers
  const candidateName = profile.name || profile.login || "Beta";
  customized.candidateName = candidateName;
  customized.topLanguage = topLang;
  customized.totalStars = stars;
  customized.publicRepos = publicRepos;
  customized.followers = followers;
  customized.forkRatioPercent = Math.round(forkRatio * 100);

  // Dynamic customization based on real stats
  if (publicRepos > 0) {
    customized.assets = [
      `${publicRepos} Public Repositories (Dahej ki list)`,
      `${stars} Total Stars (Mohallay ki izzat)`,
      `${followers} Followers vs ${following} Following (${followers > following ? 'Mashoor Shakhsiyat' : 'Sab ko follow karta phirta hai'})`,
      `Primary Tech: ${topLang} (${topLang === 'JavaScript' ? 'Nakhray baray hain' : 'Mehnat kash insaan'})`
    ];
  } else {
    customized.assets = [
      "0 Public Repositories (Ghar ka cupboard khali hai)",
      "0 Stars (Kisi ne like nahi kiya)",
      "Private repos ka bahana (Kehta hai confidential hai)",
      "Zero verifiable skills"
    ];
  }

  // Adjust score slightly based on stars and followers for uniqueness
  const bonus = Math.min(15, Math.floor(stars / 20) + Math.floor(followers / 50));
  customized.matchScore = Math.min(96, Math.max(12, customized.matchScore + (stars > 10 ? bonus : -bonus)));

  return customized;
}

// Support both browser global and Node module
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { RISHTA_FALLBACKS, generateFallbackRoast };
}
